### DISCUSSION TITLE
Help needed on `touch1200` to enter bootloader on different host platforms and MCU ports

### DISCUSSION THREAD

**Discussion by Josverl** [micropython/micropython#15306] [Core Development] [answered]
Hi,

I was pointed by @robert-hh and @andrewleech to the existence of the touch 1200 bps maner of activating a boards bootloader,
and I have tried to add support for that to mpflash by adding an option to specify the way to trigger the bootloader.
`--bootloader            -bl [manual|mpy|touch1200|none] ` 

I have a few boards that should support this in the standard builds, but I have been struggling to find how to reliably triggers this from different host platforms.
Actually I only have a working sample from Linux (Ubuntu 20.04). I have looked across github on how this is triggered , but found very different, and often quite old code, but no python code that works from windows.
(No macOs test performed yet )

| Board | Windows | Linux | Mac |
|--------|--------|--------|--------|
| SEEED_WIO_TERMINAL  | _Now Fixed_| OK | Not tested |
| ARDUINO_NANO_RP2040_CONNECT | _Now Fixed_ | OK | Not tested |

My Questions are : 
- Does anyone have a piece of sample code to trigger the 1200 BD bootloader that works from any other OS than linux 
  the different samples I found sometimes seem to fiddle with the DTR / RTS settings which I know have different behaviours between Linux and Windows 
- How can I determine if triggering the bootloader worked ? (currently I just assume so )
- How could I debug this on a target MCU ? ( ie how can I debug/ determine the bootloader state ?) 

My current code is : 
``` py
    try:
        with serial.Serial(
            port=mcu.serialport,
            baudrate=1200,
            parity=serial.PARITY_NONE,
            stopbits=serial.STOPBITS_ONE,
            bytesize=serial.EIGHTBITS,
            rtscts=True,
        ) as connection:
            print("Connection established")
            connection.rts = True
            connection.dtr = False
            time.sleep(0.4)

    except serial.SerialException as e:
        log.exception(e)
        raise MPFlashError("pySerial error: " + str(e) + "\n") from e
    except Exception as e:
        log.exception(e)
        raise MPFlashError("Error: " + str(e) + "\n") from e
```
see : [touch1200.py](https://github.com/Josverl/micropython-stubber/blob/73441cf9ee0c914bd5bba364c1d684c9c0f5e7e7/src/mpflash/mpflash/bootloader/touch1200.py#L23-L42)

**Comment by scruss** [comment, marked as answer]
This is what I use on Mac OS:

```python3
#!/usr/bin/env python3
# reset uc on given serial port by toggling rate to 1200
# -*- coding: utf-8 -*-

import sys
import time
import serial

com = serial.Serial(sys.argv[1], 1200, dsrdtr=True)
com.dtr = False
com.rts = False
time.sleep(0.2)
com.close()
time.sleep(4)
```
I nicked this code from someone on the Raspberry Pi forum, but the version I can find there is one in one of my posts.

On RP2040, you know that the reset has worked because you get a new R/w disk called **RPI-RP2** attached

**Reply by Josverl** [reply]
Thanks Stewart, 
That works on Windows with both boards above, so that is much better.
which board do you use that with ?

```
PS D:\mypython\micropython-stubber> mpflash flash -bl touch1200

22:16:58 | ℹ️ Found v1.23.0 firmware samd\SEEED_WIO_TERMINAL-v1.23.0.uf2 for SEEED_WIO_TERMINAL on COM5.
22:16:58 | ℹ️ Updating SEEED_WIO_TERMINAL on COM5 to v1.23.0
22:16:58 | ℹ️ Entering bootloader on SEEED_WIO_TERMINAL on COM5 using CDC 1200bps !
22:17:01 | ℹ️ Board is in bootloader mode
22:17:01 | ℹ️ Board ID: SAMD51P19A-WioTerminal-v0
22:17:01 | ℹ️ Copying C:\Users\josverl\Downloads\firmware\samd\SEEED_WIO_TERMINAL-v1.23.0.uf2 to E:\.
22:17:02 | ✅ Done copying, resetting the board and wait for it to restart
22:17:04 | ℹ️ Flashed 1 boards
                        Updated boards after flashing
┏━━━━━━┳━━━━━━━━━━━┳━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━┓
┃Serial┃Family     ┃Port┃Board                            ┃CPU       ┃Version┃
┡━━━━━━╇━━━━━━━━━━━╇━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━┩
│COM5  │micropython│samd│SEEED_WIO_TERMINAL               │SAMD51P19A│v1.23.0│
│      │           │    │Wio Terminal D51R with SAMD51P19A│          │       │
└──────┴───────────┴────┴─────────────────────────────────┴──────────┴───────┘
```

**Reply by scruss** [reply]
I was cheating a bit and using CircuitPython on a SparkFun Thing+ RP2040. I also use this script on Linux. Anecdotally, I'd heard that Windows did the equivalent touch1200 thing by doing `mode com1: baud=12`, but maybe it needs some CTS/DTR fiddling too.

While touch1200 can be enabled by adding [one line](https://github.com/orgs/micropython/discussions/11574#discussioncomment-5956996) to mpconfigboard.h, for every Arduino user who can't get their sketch uploaded*, there's a frustrated user whose old 1200 baud device won't work with MicroPython.

I thought that touch1200 was an mbed thing, but it seems it was first baked into the ARM-based Arduino Due USB chip, and has kind of grown from there.

*: Arduino on RP2040 must have a touch1200-enabled firmware, or it can't upload new sketches

**Reply by Josverl** [reply]
I know its only supported on a small number of boards, but now I have a decent working solution I can just try this , in addition to a `mpremote bootloader` or a manual trigger of the bootloader.

This is especially useful for 1st time flashing of boards that don't yet have micropython on them.
If 1200 touch works on the device - fine , if not, then the next try is to use the micropython activation of the bootloader. still plenty fast, and can be overridden by a user if needed.

for now I'm testing with: 
```
BL_OPTIONS = {
    "stm32": [BootloaderMethod.TOUCH_1200, BootloaderMethod.MPY, BootloaderMethod.MANUAL],
    "rp2": [BootloaderMethod.TOUCH_1200, BootloaderMethod.MPY, BootloaderMethod.MANUAL],
    "samd": [BootloaderMethod.TOUCH_1200, BootloaderMethod.MPY, BootloaderMethod.MANUAL],
    "esp32": [BootloaderMethod.NONE],
    "esp8266": [BootloaderMethod.NONE],
}
```
